src/librustc_mir/transform/add_yk_swt_calls.rs

```diff
@@ -0,0 +1,170 @@
+            new_local_decls.push(ret_val);
+
+            let crate_hash_oper = Operand::Constant(box Constant {
+                span: DUMMY_SP,
+                ty: u64_ty,
+                user_ty: None,
+                literal: ty::Const::from_u64(tcx, local_crate_hash),
+            });
+
+            let def_idx_oper = Operand::Constant(box Constant {
+                span: DUMMY_SP,
+                ty: u32_ty,
+                user_ty: None,
+                literal: ty::Const::from_u32(tcx, self.0.as_raw_u32()),
+            });
+
+            let bb_oper = Operand::Constant(box Constant {
+                span: DUMMY_SP,
+                ty: u32_ty,
+                user_ty: None,
+                literal: ty::Const::from_u32(tcx, bb.index() as u32),
+            });
+
+            let rec_fn_oper = Operand::function_handle(tcx, rec_fn_defid,
+                List::empty(), DUMMY_SP);
+
+            let term_kind = TerminatorKind::Call {
+                func: rec_fn_oper,
+                args: vec![crate_hash_oper, def_idx_oper, bb_oper],
+                destination: Some((ret_place, new_blk_idx)),
+                cleanup: None,
+                from_hir_call: false,
+            };
+
+            // Construct a new block to replace the original one.
+            let source_info = bb_data.terminator.clone().map(|t| t.source_info)
+                .or(Some(SourceInfo { span: DUMMY_SP, scope: OUTERMOST_SOURCE_SCOPE })).unwrap();
+            let replace_block = BasicBlockData {
+                statements: vec![],
+                terminator: Some(Terminator {
+                    source_info,
+                    kind: term_kind
+                }),
+                is_cleanup: false
+            };
+            replace_blocks.push(replace_block);
+        }
+
+        // Finally, commit our transformations.
+        mir.basic_blocks_mut().extend(copied_blocks);
+        mir.local_decls.extend(new_local_decls);
+        for (bb, bb_data) in replace_blocks.drain(..).enumerate() {
+            mir.basic_blocks_mut()[BasicBlock::new(bb)] = bb_data;
+        }
+    }
+}
+
+/// Given a `MirSource`, decides if we should annotate the correpsonding MIR.
+fn should_annotate(tcx: TyCtxt<'a, 'tcx, 'tcx>, src: MirSource) -> bool {
+    // Never annotate any MIR-like thing marked `#[no_trace]` or `#[naked]`. The trace record and
+    // wrapper are also marked `#[no_trace]` to prevent infinite recursion.
+    for attr in tcx.get_attrs(src.def_id).iter() {
+        if attr.check_name("no_trace") {
+            return false;
+        }
+        if attr.check_name("naked") {
+            return false;
+       }
+    }
+
+    // If there is a crate level `#![no_trace]` attribute, honour that.
+    for attr in tcx.hir.krate_attrs() {
+        if attr.check_name("no_trace") {
+            return false;
+        }
+    }
+
+    // We can't call the software tracing function if there is no libcore.
```

Comment:
This comment, and the rest of the comments in this function, state what the code does. The code is simple enough that the comments aren't very useful in that capacity. I'm wondering if we need to state *why* the code is here for these things? e.g. we might say "We can't trace a function if this crate doesn't have a link to libcore, as that's needed for the trace function"?

Reply:
OK, will review these.